### ISSUE TITLE
result object should inlcude a "fixed" field

### ISSUE THREAD

**Issue by audunolsen** [stylelint/stylelint#4664]
> What is the problem you're trying to solve?

Fix option makes me lazy and doesn't help me become a better coder, because I cannot
see what I did wrong. I want stylelint to raise my awareness of good styling practises while simultaneously speeding up my productivity.

> What solution would you like to see?

A fixed field in the result object, or alternatively have each warnings-entry have a "fixed" key.

This would allow me to have logs which may look like
```
_colors.sass
⚠ 4:7 Expected single space before "{" [fixed]
✖︎ 7:9 Expected a trailing semicolon    [fixed]
```

**Comment by audunolsen** [comment]
Here's a workaround for people interested, only drawback is that it call's the linter two times on the same file/code.

```
let [
  {results : [{warnings}]},
  {results : [{warnings: afterFix}]}    
] = await Promise.all([
  lint(options),
  lint(Object.assign(options, {fix: true}))
]);
    
for (entry of warnings) if (!afterFix
  .map(e => JSON.stringify(e))
  .includes(JSON.stringify(entry))
) entry.fixed = true;
```

<img width="580" alt="Screenshot 2020-03-31 04 15 37" src="https://user-images.githubusercontent.com/16224137/77979952-55850380-7306-11ea-8df8-0d201beda90b.png">

**Comment by jeddy3** [comment]
@audunolsen Thanks for the request, outlining your use case and using the template.

I didn't think is architecturally achievable right now, but that may change when we revisit the way rules report their fixable violations to address https://github.com/stylelint/stylelint/issues/4203 and  https://github.com/stylelint/stylelint/issues/2643. That is a sizeable piece of work that no one has found time to tackle yet.

I'll label this issue as an enhancement. It is also linked to those issues so that, hopefully, any solution to them will also factor in this use case.

**Comment by Mouvedia** [comment]
Which core formatters will have to be updated?

**Comment by ybiquitous** [comment]
> Which core [formatters](https://stylelint.io/user-guide/options#formatter) will have to be updated?

I think we can postpone changing the formatters until people request, if there are no special reasons.

**Comment by Mouvedia** [comment]
> I think we can postpone changing the formatters until people request, if there are no special reasons.

Check [OP](https://github.com/stylelint/stylelint/issues/4664#issue-590372872), he wants the boolean _and_ the output.
We should at least dogfood our internal changes.
Maybe add it to the `verbose` one first?

**Comment by ybiquitous** [comment]
> Maybe add it to the `verbose` one first?

I see the original motivation. Starting with `verbose` sounds good.

Like the following, does adding a "n problems fixed" section make sense? Can the current ESLint's behavior help us implement this change?

```sh-session
$ bin/stylelint.mjs -c scripts/visual-config.json scripts/*.css -f verbose --fix
...

5 errors found
  unit-case: 1
  string-quotes: 1
  color-no-invalid-hex: 1
  selector-max-class: 1
  unit-disallowed-list: 1

1 warning found
  custom-property-pattern: 1

2 problems fixed
  import-notation: 1
```

**Comment by Mouvedia** [comment]
> 1 problems fixed

If we did the same, we should handle plural better than them :)

**Comment by romainmenke** [comment]
I wonder if this request can be solved differently?

For example:
- disable auto-fixing in your code editor for CSS
- ensure that individual issues can be applied in code editors

This makes it possible to learn through feedback while still having access to the suggested fixes.

As far as I can tell it isn't possible to apply a single fix from Stylelint today in something like VSCode. While the same is possible with ESLint.

<img width="454" alt="Screenshot 2024-06-22 at 20 10 24" src="https://github.com/stylelint/stylelint/assets/11521496/e35e18e0-c137-43b2-bb37-f9fa1c428e74">

**Comment by Mouvedia** [comment]
I think that what you are describing is a separate and totally warranted feature.
It would require that all the fixers' signatures and `this` context be documented beforehand so that they can be called individually after an initial fix-less lint that will store the fixers and their root.
After each fix appliance that adds/removes lines, a new file lint would have to be scheduled all over again due to the configuration comments.
This sounds like a very challenging endeavour that would require the collaboration of @ota-meshi and/or @ybiquitous.
I hope I am wrong about it and that it's actually easy to pull of.

**Comment by romainmenke** [comment]
Indeed a separate feature, but one that hopefully can deliver similar value.
This issue has received no "👍"  or comments by others.
The requested feature isn't provided by other linters (as far as I can tell)

Which is why I started thinking about other way people can learn from linter feedback.

Manually applying fixes in an editor is something I do with Rust as it gives me the chance to read and understand the error while also seeing what is altered.

-------

Indeed, either exposing the `fix()` callbacks or by having the fix ranges and replacement texts.

(I wrote this suggestion before we further explored the complexities around the fix ranges and replacements texts.)

**Comment by Mouvedia** [comment]
> or by having the fix ranges and replacement texts

This would require to add a `--dry-run` flag, but once that's supported it would be really easy to achieve individual fixes indeed.

**Comment by romainmenke** [comment]
To keep https://github.com/stylelint/stylelint/pull/8267 on topic I am replying here to [your comment](https://github.com/stylelint/stylelint/pull/8267#discussion_r1901274685) @Mouvedia 

> Are you planning to eventually pass the range to `addFixData`?
> e.g.
> 
> ```
> const range = {
>   start: fixedNode.source.start,
>   end: fixedNode.source.end,
> };
> 
> addFixData({ fixersData, ruleName, range });
> ```
> 
> So that we can close #4664 since we will be able to add the range+path to the fixed section of the `verbose` formatter's output.
> i.e. click in the terminal, get to the line in your IDE and eventually scroll horizontally to reach the range


I did not intend to work on this issue, #4664

> click in the terminal, get to the line in your IDE and eventually scroll horizontally to reach the range

That would rarely work as expected. After a fix the range would have changed. `fixedNode.source.start` is the range of the original CSS, not the range after mutation. Navigating to the original position might correspond to totally different CSS after all fixes have been applied.

I honestly do not see a way how the ask from this issue can be resolved or even if we did add something how it could be sufficiently useful to be worth the effort.

I am personally in favor of not adding the requested feature.

I do understand that users might be curious about what is being fixed and why, for more informal or educational purposes but I am reluctant to expose data that could be misunderstood as generally useful or precise.

It is contradictory to both mutate a file and thereby altering all positions while also generating a report containing precise positions, positions that have been invalidated by mutating the file.

Keep in mind that multiple fixes can happen for the same range by multiple rules.
And any fix for preceding nodes also shifts the position of all subsequent nodes.

-------

I'd rather focus on the underlying issue and not on what was directly requested by this original poster.

i.e. that autofixes to quickly hide and obfuscate what was incorrect.

If a Stylelint user wants to learn more about what specific styles were incorrect and sees autofixes as a barrier to that end, then they should disable autofixes.

If a user does want autofixes while also seeing diffs, they can use existing diff tools through their choice of version control.

If a user does not want autofixes but also does not want to manually edit each warning we should focus on https://github.com/stylelint/vscode-stylelint/issues/335

------

The original poster also added their current workaround: https://github.com/stylelint/stylelint/issues/4664#issuecomment-606357438

I think this is a very valid way of approaching this issue.
This is already possible today.

I am not discounting the reason why this feature is being requested.
I just don't see how the requested implementation can work in practice.

**Comment by Mouvedia** [comment]
> `fixedNode.source.start` is the range of the original CSS, not the range after mutation.

That was just an example, my bad I meant

```
const range = {
  start: foo,
  end: bar,
}
```

I didn't delve too much into your PR because it is still a draft.

---

I was asking whether you were planning to pass `range` to `addFixData`
1. to enquiry if it was doable
2. if not, to plan for a refactor because the type of `FixerData` is absurd without `fixed` nor `range`
3. to eventually rename the function to avoid confusion
i.e because #8234 introduces `fixData` and fixedRanges which conflict semantically

**Comment by romainmenke** [comment]
Another datapoint is that ESLint completely lacks the requested feature: https://eslint.org/docs/latest/use/command-line-interface#fix-problems

I think the closest analogous feature is `--fix-dry-run` which allows you to run as if only `--fix` was set but to prevent mutating the source files.

Such a mode actually does make sense as the reported ranges would still be valid.
While the fixed output could be captured from `stdout` and reported back to the user in various ways by other tools (e.g. IDE's)

--------

> refactor because the type of `FixerData`

Maybe just a numeric counter of applied fixes per rule?

**Comment by Mouvedia** [comment]
> Maybe just a numeric counter of applied fixes per rule?

If we drop the goal of having ranges, then yes we can have something that increments in the currently named `addFixData` function. It would simplify the `verbose` formatter greatly.
i.e. if we don't plan to have anything specific stored per fix, then having `{ [ruleName: string]: number };`instead of `{ [ruleName: string]: Array<FixerData> };` would be the logical outcome

**Comment by romainmenke** [comment]
🤔 One thing I think of now is that we could record the messages.
The ranges are impossible to get right as I see things now, but the messages are just fine.

```
$ bin/stylelint.mjs -c scripts/visual-config.json scripts/*.css -f verbose --fix
...

5 errors found
  unit-case: 1
  string-quotes: 1
  color-no-invalid-hex: 1
  selector-max-class: 1
  unit-disallowed-list: 1

1 warning found
  custom-property-pattern: 1
    Expected "--foo" to be "--x-foo"
    Expected "--bar" to be "--x-bar"

1 problems fixed
  import-notation: 1
```

This could be a `Set` to only emit unique messages.

This still kinda meets the original request, to learn more about what was incorrect even when it was autofixed.

**Comment by Mouvedia** [comment]
I changed my mind on this, the `verbose` formatter can have a long output but it ought to remain concise.

e.g.
```shell
			path/to/file.css
			  1:2  ×  Unexpected foo  foo-bar
			  2:3  ×  Unexpected foo  foo-bar

			path/to/file2.css
			  3:1  ‼  Expected cat  qux-baz

			× 3 problems (2 errors, 1 warning)

			2 sources checked
			  path/to/file.css
			  path/to/file2.css

			2 errors found
			  foo-bar: 2

			1 warning found
			  qux-baz: 1

			1 problem fixed
			  color-hex-length: 1
```

As you can see, the paths are shared between the warnings, errors and fixed section.
i.e. adding ranges would have been impractical because it would to add one line per fix

---

I think that having metadata stored per fix for the unique purpose of using it in a _core_ formatter is not warranted.
Hence even if we could pass `range`, we couldn't use it.
We should create a PR once #8267 is merged which will rename the function, change the types and refactor the `verbose` formatter.

Nothing will be lost for custom formatters because

1. AFAIK you cannot match warnings with their corresponding `_postcssResult.stylelint.fixersData['foo-bar'][index]`
2. `_postcssResult` is "internal"
